dashboard/modules/log/log_agent.py

```diff
@@ -16,6 +18,208 @@
+            # E.g.,
+            #   block           = 0123\n5678\nabcde\n
+            #   lines_to_tail   = 1
+            #   num_lines       = 3
+            #
+            #   block.split(num_lines - lines_to_tail)
+            #   will produce splits:
+            #
+            #   block           = 0123\n5678\nabcde\n
+            #                     -----|-----|-------
+            #                                 ^
+            #                           where tail should start.
+            lines = block_data.split(b"\n", num_lines - lines_more)
+
+            # len(lines[0]) + 1 for the new line character split
+            nbytes_from_end += len(lines[-1])
+            logger.debug(
+                f"Found {lines_to_tail} lines from last {nbytes_from_end} bytes"
+            )
+            break
+
+        # Need to read more blocks.
+        lines_more -= num_lines
+        num_blocks_read += 1
+        nbytes_from_end += len(block_data)
+
+        logger.debug(
+            f"lines_more={lines_more}, nbytes_from_end={nbytes_from_end}, "
+            f"read_offset={read_offset}, end={offset_file_end}"
+        )
+
+        if read_offset == 0:
+            # We have read all blocks (since the start)
+            if lines_more > 0:
+                logger.debug(
+                    f"Read all {nbytes_from_end} from {file.name}, "
+                    f"but only found {lines_to_tail - lines_more} / {lines_to_tail} "
+                    "lines."
+                )
+            break
+
+        # Continuing with the previous block
+        prev_offset = read_offset
+        read_offset = max(0, read_offset - block_size)
+
+    offset_read_start = offset_file_end - nbytes_from_end
+    assert (
+        offset_read_start >= 0
+    ), f"Read start offset({offset_read_start}) should be non-negative"
+    logger.debug(
+        f"tailing {file.name}'s last {lines_to_tail} lines "
+        f"from {offset_read_start} to {offset_file_end}"
+    )
+    return offset_read_start, offset_file_end
+
+
+async def _stream_log_in_chunk(
+    context: grpc.aio.ServicerContext,
+    file: io.BufferedIOBase,
+    start_offset: int,
+    end_offset: Optional[int] = None,
+    keep_alive_interval_sec: int = -1,
+    block_size: int = BLOCK_SIZE,
+):
+    """Streaming log in chunk from start to end offset.
+
+    Stream binary file content in chunks from start offset to an end
+    offset if provided, else to the end of the file.
+
+    Args:
+        context: gRPC server side context
+        file: Binary file to stream
+        start_offset: File offset where streaming starts
+        end_offset: If None, implying streaming til the EOF.
+        keep_alive_interval_sec: Duration for which streaming will be
+            retried when reaching the file end
+        block_size: Number of bytes per chunk, exposed for testing
+
+    Return:
+        Async generator of StreamReply
+    """
+    assert "b" in file.mode, "Only binary file is supported."
```

Comment:
is it correct? what are other file formats? (is regular log also binary file?)

Reply:
we opened it in binary format for offset. Added comments.